File: packages/codex-subagent-kit/src/catalog.ts
import { existsSync, readFileSync, readdirSync } from "node:fs";
import { basename, extname, resolve } from "node:path";

import * as TOML from "@iarna/toml";

import type { AgentSpec, CatalogOptions, Category } from "./models";
import {
  CATEGORY_OVERRIDE_KEY,
  normalizeCatalogRoots,
  resolveGlobalAgentsDir,
  resolveGlobalCatalogDir,
  resolveProjectAgentsDir,
  resolveProjectCatalogDir,
} from "./paths";

export const BUILTIN_CATEGORIES_DIR = resolve(__dirname, "..", "builtin_catalog", "categories");

export const IMPORTED_AGENTS_CATEGORY: Category = {
  key: "imported-agents",
  title: "Imported & External",
  description: "Portable TOML agent definitions discovered from project/global agent directories.",
};

type CatalogState = {
  categories: Map<string, Category>;
  agents: Map<string, AgentSpec>;
};

type ParsedToml = Record<string, unknown>;

let builtinCatalogCache: CatalogState | null = null;

function readToml(path: string): ParsedToml {
  return TOML.parse(readFileSync(path, "utf8")) as ParsedToml;
}

function listDirectories(path: string): string[] {
  if (!existsSync(path)) {
    return [];
  }
  return readdirSync(path, { withFileTypes: true })
    .filter((entry) => entry.isDirectory())
    .map((entry) => resolve(path, entry.name))
    .sort();
}

function listTomlFiles(path: string): string[] {
  if (!existsSync(path)) {
    return [];
  }
  return readdirSync(path, { withFileTypes: true })
    .filter((entry) => entry.isFile() && extname(entry.name) === ".toml")
    .map((entry) => resolve(path, entry.name))
    .sort();
}

function categoryKeyFromDir(directoryName: string): string {
  const match = directoryName.match(/^(\d+)-(.*)$/);
  return match ? match[2] : directoryName;
}

function fallbackTitleFromKey(key: string): string {
  return key
    .split("-")
    .map((part) => part.charAt(0).toUpperCase() + part.slice(1))
    .join(" ");
}

export function parseCategoryDir(categoryDir: string): Category {
  const key = categoryKeyFromDir(basename(categoryDir));
  let title = fallbackTitleFromKey(key);
  let description = title;
  const readmePath = resolve(categoryDir, "README.md");

  if (!existsSync(readmePath)) {
    return { key, title, description };
  }

  const lines = readFileSync(readmePath, "utf8")
    .split(/\r?\n/)
    .map((line) => line.trim());

  for (const line of lines) {
    if (!line.startsWith("#")) {
      continue;
    }
    const parsedTitle = line.replace(/^#+\s*/, "").trim();
    const numericTitle = parsedTitle.match(/^\d+\.\s+(.*)$/);
    title = numericTitle ? numericTitle[1].trim() : parsedTitle || title;
    break;
  }

  for (const line of lines) {
    if (!line || line.startsWith("#") || line === "Included agents:") {
      continue;
    }
    description = line;
    break;
  }

  return { key, title, description };
}

function readInstructions(data: ParsedToml): string {
  const developerInstructions = data.developer_instructions;
  if (typeof developerInstructions === "string" && developerInstructions.trim()) {
    return developerInstructions.trimEnd();
  }

  const instructions = data.instructions;
  if (typeof instructions === "string" && instructions.trim()) {
    return instructions.trimEnd();
  }

  if (
    instructions &&
    typeof instructions === "object" &&
    "text" in instructions &&
    typeof instructions.text === "string" &&
    instructions.text.trim()
  ) {
    return instructions.text.trimEnd();
  }

  throw new Error("missing instructions text");
}

function requiredString(data: ParsedToml, key: string): string {
  const value = data[key];
  if (typeof value !== "string" || !value.trim()) {
    throw new Error(`missing required string field: ${key}`);
  }
  return value.trim();
}

export function parseAgentFile(
  path: string,
  source: string,
  inheritedCategory?: string,
): AgentSpec {
  const data = readToml(path);
  const explicitCategory = data[CATEGORY_OVERRIDE_KEY];

  if (
    explicitCategory !== undefined &&
    (typeof explicitCategory !== "string" || !explicitCategory.trim())
  ) {
    throw new Error(`invalid ${CATEGORY_OVERRIDE_KEY}`);
  }

  return {
    key: basename(path, ".toml"),
    category:
      typeof explicitCategory === "string" && explicitCategory.trim()
        ? explicitCategory.trim()
        : inheritedCategory ?? IMPORTED_AGENTS_CATEGORY.key,
    name: requiredString(data, "name"),
    description: requiredString(data, "description"),
    model: requiredString(data, "model"),
    reasoningEffort: requiredString(data, "model_reasoning_effort"),
    sandboxMode: requiredString(data, "sandbox_mode"),
    developerInstructions: readInstructions(data),
    source,
    definitionPath: path,
  };
}

function loadCatalogRoot(root: string, source: string): CatalogState {
  const categories = new Map<string, Category>();
  const agents = new Map<string, AgentSpec>();

  if (!existsSync(root)) {
    return { categories, agents };
  }

  for (const categoryDir of listDirectories(root)) {
    const category = parseCategoryDir(categoryDir);
    categories.set(category.key, category);
    for (const file of listTomlFiles(categoryDir)) {
      const agent = parseAgentFile(file, source, category.key);
      agents.set(agent.key, agent);
    }
  }

  return { categories, agents };
}

function getBuiltinCatalog(): CatalogState {
  if (builtinCatalogCache) {
    return builtinCatalogCache;
  }

  const builtin = loadCatalogRoot(BUILTIN_CATEGORIES_DIR, "builtin");
  if (builtin.categories.size === 0) {
    throw new Error(`builtin catalog is empty: ${BUILTIN_CATEGORIES_DIR}`);
  }
  builtinCatalogCache = builtin;
  return builtin;
}

function mergeCatalogInto(target: CatalogState, incoming: CatalogState): void {
  for (const [key, category] of incoming.categories) {
    target.categories.set(key, category);
  }
  for (const [key, agent] of incoming.agents) {
    target.agents.set(key, agent);
  }
}

function loadExternalAgents(
  directory: string,
  source: string,
  inheritedAgents: Map<string, AgentSpec>,
): AgentSpec[] {
  if (!existsSync(directory)) {
    return [];
  }

  const loaded: AgentSpec[] = [];
  for (const file of listTomlFiles(directory)) {
    try {
      const inheritedCategory = inheritedAgents.get(basename(file, ".toml"))?.category;
      loaded.push(parseAgentFile(file, source, inheritedCategory));
    } catch {
      continue;
    }
  }
  return loaded;
}

function buildCatalogState(options: CatalogOptions = {}): CatalogState {
  const includeProject = options.includeProject ?? false;
  const includeGlobal = options.includeGlobal ?? false;
  const homeDir = options.homeDir;
  const catalogRoots = normalizeCatalogRoots(options.catalogRoots);

  const builtin = getBuiltinCatalog();
  const state: CatalogState = {
    categories: new Map(builtin.categories),
    agents: new Map(builtin.agents),
  };

  if (includeGlobal) {
    mergeCatalogInto(state, loadCatalogRoot(resolveGlobalCatalogDir(homeDir), "global-catalog"));
  }

  if (includeProject && options.projectRoot) {
    mergeCatalogInto(
      state,
      loadCatalogRoot(resolveProjectCatalogDir(options.projectRoot), "project-catalog"),
    );
  }

  for (const root of catalogRoots) {
    mergeCatalogInto(state, loadCatalogRoot(root, "catalog-root"));
  }

  state.categories.set(IMPORTED_AGENTS_CATEGORY.key, IMPORTED_AGENTS_CATEGORY);

  if (includeGlobal) {
    for (const agent of loadExternalAgents(
      resolveGlobalAgentsDir(homeDir),
      "global",
      state.agents,
    )) {
      state.agents.set(agent.key, agent);
    }
  }

  if (includeProject && options.projectRoot) {
    for (const agent of loadExternalAgents(
      resolveProjectAgentsDir(options.projectRoot),
      "project",
      state.agents,
    )) {
      state.agents.set(agent.key, agent);
    }
  }

  return state;
}

export function getAgentMap(options: CatalogOptions = {}): Map<string, AgentSpec> {
  return buildCatalogState(options).agents;
}

export function getCategories(options: CatalogOptions = {}): Category[] {
  const state = buildCatalogState(options);
  const usedCategories = new Set([...state.agents.values()].map((agent) => agent.category));
  return [...state.categories.values()].filter((category) => usedCategories.has(category.key));
}

export function getAgents(options: CatalogOptions = {}): AgentSpec[] {
  const categories = getCategories(options);
  const agents = [...getAgentMap(options).values()];
  const categoryOrder = new Map(categories.map((category, index) => [category.key, index]));

  return agents.sort((left, right) => {
    const leftOrder = categoryOrder.get(left.category) ?? categoryOrder.size;
    const rightOrder = categoryOrder.get(right.category) ?? categoryOrder.size;
    if (leftOrder !== rightOrder) {
      return leftOrder - rightOrder;
    }
    const byName = left.name.localeCompare(right.name, undefined, { sensitivity: "base" });
    if (byName !== 0) {
      return byName;
    }
    return left.key.localeCompare(right.key);
  });
}

export function getAgentsByCategory(
  categoryKeys: Set<string> | undefined,
  options: CatalogOptions = {},
): AgentSpec[] {
  const agents = getAgents(options);
  if (!categoryKeys || categoryKeys.size === 0) {
    return agents;
  }
  return agents.filter((agent) => categoryKeys.has(agent.category));
}

export function renderCatalogOutput(options: CatalogOptions = {}): string {
  const categories = getCategories(options);
  const agents = getAgents(options);
  const lines: string[] = [];

  for (const category of categories) {
    lines.push(`[${category.title}]`);
    lines.push(`  key: ${category.key}`);
    lines.push(`  description: ${category.description}`);

    for (const agent of agents) {
      if (agent.category !== category.key) {
        continue;
      }
      const sourceSuffix = agent.source === "builtin" ? "" : ` [${agent.source}]`;
      lines.push(`  - ${agent.key}: ${agent.description}${sourceSuffix}`);
    }

    lines.push("");
  }

  return lines.join("\n").trimEnd();
}

File: src/codex_subagent_kit/catalog.py
from __future__ import annotations

import tomllib
from functools import lru_cache
from pathlib import Path

from .app_paths import (
    CATEGORY_OVERRIDE_KEY,
    resolve_global_catalog_dir,
    resolve_global_source_categories_dirs,
    resolve_project_catalog_dir,
    resolve_project_source_categories_dirs,
)
from .models import AgentSpec, Category


BUILTIN_CATEGORIES_DIR = Path(__file__).resolve().parent / "builtin_catalog" / "categories"
IMPORTED_AGENTS_CATEGORY = Category(
    key="imported-agents",
    title="Imported & External",
    description="Portable TOML agent definitions discovered from project/global agent directories.",
)


def resolve_project_agents_dir(project_root: Path) -> Path:
    return project_root / ".codex" / "agents"


def resolve_global_agents_dir() -> Path:
    return Path.home() / ".codex" / "agents"


def normalize_catalog_roots(catalog_roots: tuple[Path, ...] | None = None) -> tuple[Path, ...]:
    if not catalog_roots:
        return ()
    return tuple(root.resolve() for root in catalog_roots)


def _category_key_from_dir(directory_name: str) -> str:
    prefix, separator, remainder = directory_name.partition("-")
    if separator and prefix.isdigit():
        return remainder
    return directory_name


def _fallback_title_from_key(key: str) -> str:
    return " ".join(part.capitalize() for part in key.split("-"))


def _parse_category_dir(category_dir: Path) -> Category:
    key = _category_key_from_dir(category_dir.name)
    title = _fallback_title_from_key(key)
    description = title
    readme_path = category_dir / "README.md"
    if not readme_path.exists():
        return Category(key=key, title=title, description=description)

    lines = [line.strip() for line in readme_path.read_text(encoding="utf-8").splitlines()]
    for line in lines:
        if line.startswith("#"):
            parsed_title = line.lstrip("#").strip()
            numeric_prefix, separator, remainder = parsed_title.partition(". ")
            if separator and numeric_prefix.isdigit():
                title = remainder.strip()
            elif parsed_title:
                title = parsed_title
            break

    for line in lines:
        if not line or line.startswith("#") or line == "Included agents:":
            continue
        description = line
        break

    return Category(key=key, title=title, description=description)


def _parse_agent_file(
    path: Path,
    *,
    inherited_category: str | None = None,
    source: str,
) -> AgentSpec:
    with path.open("rb") as handle:
        data = tomllib.load(handle)

    instructions: str | None = None
    developer_instructions = data.get("developer_instructions")
    if isinstance(developer_instructions, str):
        instructions = developer_instructions
    else:
        raw_instructions = data.get("instructions")
        if isinstance(raw_instructions, str):
            instructions = raw_instructions
        elif isinstance(raw_instructions, dict):
            nested_text = raw_instructions.get("text")
            instructions = nested_text if isinstance(nested_text, str) else None
    if not isinstance(instructions, str) or not instructions.strip():
        raise ValueError("missing instructions text")

    name = data.get("name")
    description = data.get("description")
    model = data.get("model")
    reasoning_effort = data.get("model_reasoning_effort")
    sandbox_mode = data.get("sandbox_mode")
    if not all(
        isinstance(value, str) and value.strip()
        for value in (name, description, model, reasoning_effort, sandbox_mode)
    ):
        raise ValueError("missing required string fields")

    explicit_category = data.get(CATEGORY_OVERRIDE_KEY)
    if explicit_category is not None and (not isinstance(explicit_category, str) or not explicit_category.strip()):
        raise ValueError(f"invalid {CATEGORY_OVERRIDE_KEY}")

    category = (
        explicit_category.strip()
        if isinstance(explicit_category, str) and explicit_category.strip()
        else inherited_category or IMPORTED_AGENTS_CATEGORY.key
    )
    return AgentSpec(
        key=path.stem,
        category=category,
        name=name.strip(),
        description=description.strip(),
        model=model.strip(),
        reasoning_effort=reasoning_effort.strip(),
        sandbox_mode=sandbox_mode.strip(),
        developer_instructions=instructions.rstrip(),
        source=source,
        definition_path=path,
    )


def _load_catalog_root(
    root: Path,
    *,
    source: str,
) -> tuple[dict[str, Category], dict[str, AgentSpec]]:
    if not root.exists():
        return {}, {}

    categories: dict[str, Category] = {}
    agents: dict[str, AgentSpec] = {}
    for category_dir in sorted(root.iterdir()):
        if not category_dir.is_dir():
            continue
        category = _parse_category_dir(category_dir)
        categories[category.key] = category
        for path in sorted(category_dir.glob("*.toml")):
            agent = _parse_agent_file(path, inherited_category=category.key, source=source)
            agents[agent.key] = agent
    return categories, agents


@lru_cache(maxsize=1)
def _get_builtin_catalog() -> tuple[dict[str, Category], dict[str, AgentSpec]]:
    categories, agents = _load_catalog_root(BUILTIN_CATEGORIES_DIR, source="builtin")
    if not categories:
        raise RuntimeError(f"builtin catalog is empty: {BUILTIN_CATEGORIES_DIR}")
    return categories, agents


def _load_external_agents(
    *,
    directory: Path,
    source: str,
    inherited_agents: dict[str, AgentSpec],
) -> list[AgentSpec]:
    if not directory.exists():
        return []

    loaded: list[AgentSpec] = []
    for path in sorted(directory.glob("*.toml")):
        inherited_category = inherited_agents.get(path.stem).category if path.stem in inherited_agents else None
        try:
            loaded.append(_parse_agent_file(path, inherited_category=inherited_category, source=source))
        except (OSError, ValueError, tomllib.TOMLDecodeError):
            continue
    return loaded


def _merge_catalog_roots(
    *,
    project_root: Path | None,
    include_project: bool,
    include_global: bool,
    catalog_roots: tuple[Path, ...],
) -> tuple[dict[str, Category], dict[str, AgentSpec]]:
    builtin_categories, builtin_agents = _get_builtin_catalog()
    categories: dict[str, Category] = dict(builtin_categories)
    agents: dict[str, AgentSpec] = dict(builtin_agents)

    if include_global:
        for root in resolve_global_source_categories_dirs():
            extra_categories, extra_agents = _load_catalog_root(
                root,
                source=f"global-source:{root.parent.name}",
            )
            categories.update(extra_categories)
            agents.update(extra_agents)

        extra_categories, extra_agents = _load_catalog_root(
            resolve_global_catalog_dir(),
            source="global-catalog",
        )
        categories.update(extra_categories)
        agents.update(extra_agents)

    if include_project and project_root is not None:
        for root in resolve_project_source_categories_dirs(project_root):
            extra_categories, extra_agents = _load_catalog_root(
                root,
                source=f"project-source:{root.parent.name}",
            )
            categories.update(extra_categories)
            agents.update(extra_agents)

        extra_categories, extra_agents = _load_catalog_root(
            resolve_project_catalog_dir(project_root),
            source="project-catalog",
        )
        categories.update(extra_categories)
        agents.update(extra_agents)

    for root in catalog_roots:
        extra_categories, extra_agents = _load_catalog_root(root, source="catalog-root")
        categories.update(extra_categories)
        agents.update(extra_agents)

    categories[IMPORTED_AGENTS_CATEGORY.key] = IMPORTED_AGENTS_CATEGORY
    return categories, agents


def get_agent_map(
    *,
    project_root: Path | None = None,
    include_project: bool = False,
    include_global: bool = False,
    catalog_roots: tuple[Path, ...] | None = None,
) -> dict[str, AgentSpec]:
    normalized_catalog_roots = normalize_catalog_roots(catalog_roots)
    categories, agent_map = _merge_catalog_roots(
        project_root=project_root,
        include_project=include_project,
        include_global=include_global,
        catalog_roots=normalized_catalog_roots,
    )
    del categories

    if include_global:
        for agent in _load_external_agents(
            directory=resolve_global_agents_dir(),
            source="global",
            inherited_agents=agent_map,
        ):
            agent_map[agent.key] = agent

    if include_project and project_root is not None:
        for agent in _load_external_agents(
            directory=resolve_project_agents_dir(project_root),
            source="project",
            inherited_agents=agent_map,
        ):
            agent_map[agent.key] = agent

    return agent_map


def get_agents(
    *,
    project_root: Path | None = None,
    include_project: bool = False,
    include_global: bool = False,
    catalog_roots: tuple[Path, ...] | None = None,
) -> tuple[AgentSpec, ...]:
    normalized_catalog_roots = normalize_catalog_roots(catalog_roots)
    categories = get_categories(
        project_root=project_root,
        include_project=include_project,
        include_global=include_global,
        catalog_roots=normalized_catalog_roots,
    )
    agent_map = get_agent_map(
        project_root=project_root,
        include_project=include_project,
        include_global=include_global,
        catalog_roots=normalized_catalog_roots,
    )
    category_order = {category.key: index for index, category in enumerate(categories)}
    return tuple(
        sorted(
            agent_map.values(),
            key=lambda agent: (
                category_order.get(agent.category, len(category_order)),
                agent.name.lower(),
                agent.key,
            ),
        )
    )


def get_categories(
    *,
    project_root: Path | None = None,
    include_project: bool = False,
    include_global: bool = False,
    catalog_roots: tuple[Path, ...] | None = None,
) -> tuple[Category, ...]:
    normalized_catalog_roots = normalize_catalog_roots(catalog_roots)
    categories, _ = _merge_catalog_roots(
        project_root=project_root,
        include_project=include_project,
        include_global=include_global,
        catalog_roots=normalized_catalog_roots,
    )
    used_categories = {
        agent.category
        for agent in get_agent_map(
            project_root=project_root,
            include_project=include_project,
            include_global=include_global,
            catalog_roots=normalized_catalog_roots,
        ).values()
    }
    return tuple(category for key, category in categories.items() if key in used_categories)


def get_agents_by_category(
    category_keys: set[str] | None = None,
    *,
    project_root: Path | None = None,
    include_project: bool = False,
    include_global: bool = False,
    catalog_roots: tuple[Path, ...] | None = None,
) -> list[AgentSpec]:
    agents = list(
        get_agents(
            project_root=project_root,
            include_project=include_project,
            include_global=include_global,
            catalog_roots=catalog_roots,
        )
    )
    if not category_keys:
        return agents
    return [agent for agent in agents if agent.category in category_keys]
